### PR TITLE
Default config - Add new e2e tests for the cht-form web component

### PR DESCRIPTION
**Pull request by tatilepizs**
# Description

This PR adds new automated tests for the cht-form web component. It covers all the available forms for the `default` config:

- [x] death_report
- [x] delivery
- [x] pnc_danger_sign_follow_up_baby
- [x] pnc_dange_sign_follow_up_mother
- [x] pregnancy_danger_sign_follow_up
- [x] pregnancy_danger_sign
- [x] pregnancy_facility_visit_reminder
- [x] pregnancy_home_visit
- [x] pregnancy
- [x] replace_user
- [x] undo_death_report
- [x] enketo_widgets
- [x] person-create
- [x] person-edit
- [x] language test (included in death report)

medic/cht-core#[number]

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:cht-form-e2e-tests/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:cht-form-e2e-tests/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:cht-form-e2e-tests/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by jkuester** [comment]
@tatilepizs , I had a change to look into those enhancements we had briefly discussed awhile back that would basically be the final pieces needed to make these tests "fully automated"!  Since I pretty much just had to write the code to figure out what was going to actually work, I just created a new PR for adding the changes to this branch (so I don't have to try and put it all in a review commit here...)  https://github.com/medic/cht-core/pull/8647

If those changes look good to you, feel free to just pull them in here!

File: tests/e2e/cht-form/mock-config.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+const generateFormData = async (formPath) => {
+  const formXml = await fs.readFile(formPath, 'utf-8');
+  // This is essentially the same code used by the test-harness to generate the form html and model.
+  // If this code changes, the test-harness will need to be updated as well.
+  const { form: formHtml, model: formModel } = await generateXformService.generate(formXml);
+  return { formHtml, formModel, formXml };
+};
+
+const startMockApp = async (config, formType, formName) => {
```
Comment (m5r) [inline tests/e2e/cht-form/mock-config.js:27]: :warning: please don't implement this change, I'm just thinking out loud for potential improvements :warning: 

I would love if we could write e2e test files in typescript. Ideally these parameters would have dynamic types that map to the actual configs and forms available
```ts
mockConfig.startMockApp('default', 'app', 'pregnancy');
//                                        ^ autocompletes to 'death_report' | 'delivery' | 'pregnancy'... and so on when the first arg is 'default'
//                      ^ autocompletes to 'default' | 'standard' | 'covid-19' | 'demo'
```

**Review by m5r — APPROVED** [review]
Great job, I left a few small suggestions but besides those it's good to go!  
I tried re-running the CI a couple times but it doesn't seem to pass. Are we introducing some flaky tests or are those real failures ? :sweat_smile:

File: tests/e2e/cht-form/default/delivery.wdio-spec.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    await deliveryForm.fillWomanDeathInformation({
+      date: DATE,
+      place: 'health_facility',
+      notes: 'Test notes - Mother\'s death'
+    });
+    await genericForm.nextPage();
+    await genericForm.nextPage();
+
+    const data = await mockConfig.submitForm();
+    const jsonObj = data[1];
```
Comment (jkuester) [inline tests/e2e/cht-form/default/delivery.wdio-spec.js:120]: Can you also assert the contents of `data[0]` (the actual delivery form)?  With these forms that are resulting in multiple docs, I think we want to be extra clear about what we expect to be in each of these docs.

File: tests/e2e/cht-form/default/death-report.wdio-spec.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    await deathReportForm.setDeathInformation(deathNote);
+    await deathReportForm.setDeathDate(date);
+    await genericForm.nextPage();
+
+    const {deathDate, deathInformation} = await deathReportForm.getSummaryDetails();
+    expect(deathDate).to.equal(date);
+    expect(deathInformation).to.equal(deathNote);
+
+    const data = await mockConfig.submitForm();
+    const jsonObj = data[0].fields.death_details;
```
Comment (jkuester) [inline tests/e2e/cht-form/default/death-report.wdio-spec.js:31]:
```suggestion
    const [doc, ...additionalDocs] = await mockConfig.submitForm();
    expect(additionalDocs).to.be.empty;
    const jsonObj = doc.fields.death_details;
```

These tests are well positioned to be able to assert exactly what docs are created by a form (something that is more challenging in a regular e2e test).  Could you possibly update similar code in tests like this to assert that only one doc is created?

**Review by jkuester — APPROVED** [review]
Super great!  I really appreciate the consistancy of the code and thoroughness of the assertions!  

I did update the GitHub Actions config in my `7462_forms_module` branch to hopefully fix the CI failures.

File: tests/e2e/cht-form/mock-config.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    server = mockApp.listen(resolve);
+  }).then(async () => {
+    await browser.url(`http://localhost:${server.address().port}`);
+    await browser.execute((formData) => {
+      const myForm = document.getElementById('myform');
+      myForm.formHtml = formData.formHtml;
+      myForm.formModel = formData.formModel;
+      myForm.formXml = formData.formXml;
+    }, formData);
+  });
```
[the remark this answers is not in the document]
Comment (tatilepizs) [reply tests/e2e/cht-form/mock-config.js]: Thank you! Done ⭐

File: tests/e2e/cht-form/wdio.conf.js [hunk 5]
```diff
@@ -1,198 +1,22 @@
...
-  // then the current working directory is where your `package.json` resides, so `wdio`
-  // will be called from there.
-  //
+const mockConfig = require('./mock-config');
 
-  suites: {
-    all: ['**/*.wdio-spec.js']
-  },
+// Override specific properties from wdio base config
+const defaultConfig = Object.assign(wdioBaseConfig.config, {
```
[the remark this answers is not in the document]
Comment (tatilepizs) [reply tests/e2e/cht-form/wdio.conf.js:14]: Done 😄

File: tests/e2e/cht-form/mock-config.js (hunk 1 above)
Comment (tatilepizs) [reply tests/e2e/cht-form/mock-config.js:27]: This is something that Josh and I had talked about on different occasions, it is a great idea, and it is something that we should do also for the e2e tests in `cht-core`

File: tests/page-objects/default/enketo/death-report.page.js [hunk 6]
```diff
@@ -52,7 +52,18 @@ const submitDeathReport = async ({
...
+  return {
+    details: await $('span[data-itext-id="/death_report/death_details:label"].active').getText(),
+    date: await $('span[data-itext-id="/death_report/death_details/date_of_death:label"].active').getText(),
+    place: await $('span[data-itext-id="/death_report/death_details/place_of_death:label"].active').getText(),
+    healthFacility: await deathPlace(PLACE_OF_DEATH.healthFacility).nextElement().getText(),
+    home: await deathPlace(PLACE_OF_DEATH.home).nextElement().getText(),
+    other: await deathPlace(PLACE_OF_DEATH.other).nextElement().getText(),
+    notes: await $('span[data-itext-id="/death_report/death_details/death_information:label"].active').getText()
+  };
 };
```
[the remark this answers is not in the document]
Comment (tatilepizs) [reply tests/page-objects/default/enketo/death-report.page.js:67]: 🤯 you are completely right, I will update it, thank you! ⭐

File: tests/e2e/cht-form/default/delivery.wdio-spec.js (hunk 2 above)
Comment (tatilepizs) [reply tests/e2e/cht-form/default/delivery.wdio-spec.js:120]: Of course. Done 😄

File: tests/e2e/cht-form/default/death-report.wdio-spec.js (hunk 3 above)
Comment (tatilepizs) [reply tests/e2e/cht-form/default/death-report.wdio-spec.js:31]: Sure, that is a great idea.
Done ⭐

**Pull request merged**